src/kakarot/instructions/system_operations.cairo

```diff
@@ -467,10 +538,74 @@ namespace CreateHelper {
             bytecode=ctx.return_data,
         );
         local ctx: model.ExecutionContext* = ExecutionContext.update_sub_context(ctx.calling_context, ctx);
+        
+        // Append contracts to selfdestruct to the calling_context
+        Helpers.fill_array(
+            fill_len=ctx.sub_context.destroy_contracts_len,
+            input_arr=ctx.sub_context.destroy_contracts,
+            output_arr=ctx.destroy_contracts + ctx.destroy_contracts_len,
+        );
+        let ctx = ExecutionContext.update_destroy_contracts(
+            self=ctx,
+            destroy_contracts_len=ctx.destroy_contracts_len + ctx.sub_context.destroy_contracts_len,
+            destroy_contracts=ctx.destroy_contracts,
+            stop=ctx.stopped,
+        );
+
         let (address_high, address_low) = split_felt(ctx.sub_context.evm_contract_address);
         let stack = Stack.push(ctx.stack, Uint256(low=address_low, high=address_high));
         let ctx = ExecutionContext.update_stack(ctx, stack);
 
         return ctx;
     }
 }
+
+namespace SelfDestructHelper {
+
+    // @notice helper for SELFDESTRUCT operation.
+    // @notice overwrite contract account bytecode with 0s
+    // @notice remove contract from registry
+    // @return The pointer to the updated execution_context.
+    func selfdestruct{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        alloc_locals;
+
+        if (ctx.destroy_contracts_len == 0) {
+            return (ctx);
+        }
+
+        // Get bytecode_len from the EVM contract
+        let starknet_contract_address = [ctx.destroy_contracts];
+        let (bytecode_len) = IEvmContract.bytecode_len(
+            contract_address=starknet_contract_address
+        );
+        let (erase_data) = alloc();
+        Helpers.fill(bytecode_len, erase_data, 0);
+        IEvmContract.write_bytecode(
+            contract_address=starknet_contract_address,
+            bytecode_len=0,
+            bytecode=erase_data
+        );
+
+        // Remove contract from registry
+        let (registry_address_) = registry_address.read();
+        IRegistry.set_account_entry(
+            contract_address=registry_address_,
+            starknet_contract_address=starknet_contract_address,
+            evm_contract_address=0
+        );
+
+        let ctx = ExecutionContext.update_destroy_contracts(
```

Comment:
you should maybe create a loop with a dedicated internal function to avoid creating new ExecutionContext each time, and consequently save some steps.

That being said, it's smooth like that and I don't expect selfdestruct to be called dozens of time so it's probably fine like that

Reply:
Good idea, yes maybe not necessary here but I will keep it in mind for further implementations

Reply:
Added it finally